cmd/controller/main.go

```diff
@@ -155,21 +156,31 @@ func main() {
 		Keychain:       kpackKeychain,
 	}
 
+	lifecycleProvider, err := lifecycle.NewProvider(*lifecycleImage, &registry.Client{}, kpackKeychain)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	configMapWatcher.Watch(lifecycle.ConfigName, lifecycleProvider.UpdateImage)
+
 	builderCreator := &cnb.RemoteBuilderCreator{
 		RegistryClient:         &registry.Client{},
-		LifecycleImage:         *lifecycleImage,
 		KpackVersion:           cmd.Identifer,
+		LifecycleProvider:      lifecycleProvider,
 		NewBuildpackRepository: newBuildpackRepository(kpackKeychain),
 	}
 
 	buildController := build.NewController(options, k8sClient, buildInformer, podInformer, metadataRetriever, buildpodGenerator)
 	imageController := image.NewController(options, k8sClient, imageInformer, buildInformer, duckBuilderInformer, sourceResolverInformer, pvcInformer)
 	sourceResolverController := sourceresolver.NewController(options, sourceResolverInformer, gitResolver, blobResolver, registryResolver)
-	builderController := builder.NewController(options, builderInformer, builderCreator, keychainFactory, clusterStoreInformer, clusterStackInformer)
-	clusterBuilderController := clusterBuilder.NewController(options, clusterBuilderInformer, builderCreator, keychainFactory, clusterStoreInformer, clusterStackInformer)
+	builderController, builderResync := builder.NewController(options, builderInformer, builderCreator, keychainFactory, clusterStoreInformer, clusterStackInformer)
+	clusterBuilderController, clusterBuilderResync := clusterBuilder.NewController(options, clusterBuilderInformer, builderCreator, keychainFactory, clusterStoreInformer, clusterStackInformer)
 	clusterStoreController := clusterstore.NewController(options, clusterStoreInformer, remoteStoreReader)
 	clusterStackController := clusterstack.NewController(options, clusterStackInformer, remoteStackReader)
 
+	lifecycleProvider.RegisterCallback(builderResync)
+	lifecycleProvider.RegisterCallback(clusterBuilderResync)
```

Comment:
There's something a little unnerving seeing callbacks like this. I'm wondering if listening for messages on a channel is idiomatic way to do this in Golang..

Reply:
Perhaps, `callback` is not the correct term. 

Registering event handlers is pretty ubiquitous in the k8s/go ecosystem ([for example](https://github.com/pivotal/kpack/blob/46fe4fc3a99696d451a9e2fef2902451dd120108/pkg/reconciler/build/build.go#L52-L57)).  Channel listening seems like it would introduce unnecessary concurrency.

Reply:
In particular, it's the fact that we're passing the following around for another class to execute:

```go
	return impl, func() {
		impl.GlobalResync(builderInformer.Informer())
	}
```

Perhaps we've discussed this before, but it doesn't feel very open/closed. Just wondering if there's a different way to express this.